### DISCUSSION TITLE
Recommended way to use FastAPI with Firebase (Pyrebase4  or firebase_admin)

### DISCUSSION THREAD

**Discussion by JuniorGunner** [fastapi/fastapi#6962] [Questions] [answered]
I want to know how to setup the comunication between FastAPI and the Firebase data storage (Firestore) through Pyrebase4. I didn't find any info or example  of this case.

**Comment by krishnardt** [comment]
Pyrebase is pretty straight forward to use/understand than firebase-admin in implementation.
I tried both of them.
Pyrebase has realtime streaming/listening function.
In older versions of Firebase_admin there is no realtime listener for python SDK. It is implemented in the latest versions.

Currently  in feature-wise, there is no much difference between both libraries.

**Comment by JuniorGunner** [comment]
Do you have some example of basic setup of FastAPI with Pyrebase4? I mean, some simple CRUD using it. I couldn't find any.

**Comment by krishnardt** [comment]
You can check this sample..

```
import pyrebase

fireBaseConfig = {
        
    'apiKey': "******************************",
    'authDomain': "********************************.firebaseapp.com",
    'databaseURL': "https://**************.firebaseio.com",
    'projectId': "**************",
    'storageBucket': "**************.appspot.com",
    'messagingSenderId': "*****************",
    'appId': "**************************",
    'measurementId': "**************"
        }
```

we will have to get those firebaseConfig details by registering in firebase.



```
firebase = pyrebase.initialize_app(fireBaseConfig)
auth = firebase.auth()
db = firebase.database()
```

you can save above db config details in seperate file and can import them.


```
import json
from fastapi import FastAPI

app=FastAPI()

@app.get("/")
async def get_data():
	"""to get data from a given table"""
	some_data = db.child('any_table_name').get()
	print(some_data)
	return some_data

@app.get("/users")
async def get_users():
	"""to get users details"""
	return db.child('users').get()

@app.get("/info")
async def get_info():
	"""to get every detail in the database."""
	return db.child().get()
```

for reference 
[https://github.com/nhorvath/Pyrebase4](url)

**Comment by JuniorGunner** [comment]
I did the following, but for a reason that I didn't find yet, I can't reach the firebase data.

![image](https://user-images.githubusercontent.com/12654382/104919667-60805180-5975-11eb-82c1-65a277e8d664.png)

I got http 200 but without  the Firebase data. Any tip about that?

**Comment by krishnardt** [comment]
have you created api_key etc., in firebase config details and used in api?

need more info

**Comment by JuniorGunner** [comment]
My project structure:

my_api
 |_app
      |_config
      |     |_database.py
      |_routes
      |      |_category.py
      |_schemas
      |        |_category.py
      |_main.py


My app/config/database.py:

![image](https://user-images.githubusercontent.com/12654382/104932115-0b007080-5986-11eb-8d8c-5aa83381b2a8.png)

My app/schemas/category.py:

![image](https://user-images.githubusercontent.com/12654382/104932335-4438e080-5986-11eb-9ee1-ad1f4de1e20b.png)

My app/routes/category.py:

![image](https://user-images.githubusercontent.com/12654382/104932534-7e09e700-5986-11eb-9174-04bcf9e3a037.png)


My app/main.py:

![image](https://user-images.githubusercontent.com/12654382/104932670-abef2b80-5986-11eb-8372-8ca2f2147d71.png)

Am I missing something?

**Comment by JuniorGunner** [comment]
![image](https://user-images.githubusercontent.com/12654382/104957440-ea97dc80-59ac-11eb-9ed0-4f0b3eaa56b0.png)

Just for confirmation, I have one single record saved for test.

**Comment by krishnardt** [comment]
Can you check in realtime database?
Currently pyrebase doesn't have support to cloud firestore. It has support for realtime database.
Whatever you do transactions with pyrebase, they will be stored and fetched from realtime database.

If you need to work on firestore,use firebase_admin sdk.

follow this link for reference...
https://github.com/thisbejim/Pyrebase/issues/268#issuecomment-545557451

**Comment by JuniorGunner** [comment]
Thank you for your help. Since it doesn't have support I'll change to realtime database or firebase_admin sdk.

**Comment by DotSyndicate** [comment, marked as answer]
@JuniorGunner, I hope you made it work. I am also in a similar situation. If your app worked, can you please share your code with me? Like, I need to understand how you made it work. Thank you. You can make it as an answer to [this](https://github.com/tiangolo/fastapi/issues/3508#issue-940435846).

**Reply by SyedAzeemJaved** [reply]
Do we have any update on this? I too am facing a similar issue, where I want to initialize the `firebase` instance only once, also how would it be handling multiple users of my API trying to upload/fetch data? Would the requests be blocking in this case? That is while the instance is being 'used', other requests from other concurrent users have to wait?

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue 👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by emrothenberg** [comment]
You can use empyrebase (https://github.com/emrothenberg/empyrebase), it's a fork of pyrebase4 and supports Firestore.
